iroh-gateway/src/core.rs

```diff
@@ -0,0 +1,412 @@
+}
+
+async fn serve_html(
+    req: &Request,
+    client: Client,
+    headers: &mut HashMap<String, String>,
+) -> Result<GatewayResponse, GatewayError> {
+    let body = client
+        .get_file(format!("{}", req.full_content_path).as_str())
+        .await;
+    let body = match body {
+        Ok(b) => b,
+        Err(e) => {
+            let msg = format!("{}", e);
+            return error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+        }
+    };
+    response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+}
+
+async fn serve_fs(
+    req: &Request,
+    client: Client,
+    headers: &mut HashMap<String, String>,
+) -> Result<GatewayResponse, GatewayError> {
+    let body = client
+        .get_file(format!("{}", req.full_content_path).as_str())
+        .await;
+    let body = match body {
+        Ok(b) => b,
+        Err(e) => {
+            let msg = format!("{}", e);
+            return error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+        }
+    };
+    let name = add_content_disposition_headers(
+        headers,
+        &req.query_file_name,
+        &req.content_path,
+        req.download,
+    );
+    add_content_type_headers(headers, &name);
+    response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+}
+
+fn add_user_headers(headers: &mut HashMap<String, String>, user_headers: HashMap<String, String>) {
+    for (key, value) in user_headers {
+        headers.insert(key, value);
+    }
+}
+
+fn add_content_type_headers(headers: &mut HashMap<String, String>, name: &str) {
+    let guess = mime_guess::from_path(name);
+    let content_type = guess.first_or_octet_stream().to_string();
+    headers.insert(CONTENT_TYPE.to_string(), content_type);
+}
+
+fn add_content_disposition_headers(
+    headers: &mut HashMap<String, String>,
+    filename: &str,
+    content_path: &str,
+    download: bool,
+) -> String {
+    let mut name = get_filename(content_path);
+    if filename != "" {
+        let disposition = if download {
+            DISPOSITION_ATTACHMENT
+        } else {
+            DISPOSITION_INLINE
+        };
+        name = filename.to_string();
+        set_content_disposition_headers(headers, filename, disposition);
+    }
+    name
+}
+
+fn set_content_disposition_headers(
+    headers: &mut HashMap<String, String>,
+    filename: &str,
+    disposition: &str,
+) {
+    headers.insert(
+        CONTENT_DISPOSITION.to_string(),
+        format!("{}; filename={}", disposition, filename),
+    );
+}
+
+fn get_filename(content_path: &str) -> String {
+    content_path
+        .split("/")
+        .filter(|s| !s.is_empty())
+        .map(|s| s.to_string())
+        .last()
+        .unwrap_or_default()
+}
+
+fn response(
+    status_code: StatusCode,
+    body: BoxBody,
+    headers: &HashMap<String, String>,
+) -> Result<GatewayResponse, GatewayError> {
+    Ok(GatewayResponse {
+        status_code,
+        body: body,
+        headers: headers.clone(),
```

Comment:
instead of `clone`ing here, just pass the owned value as the argument, so that the caller can decide about it